Add a sqlite database browser

tttpeng opened the pull request:
Hi,
I've added a database browser for opening a .db file.
Sometimes, Some people may want to view SQLite databases on device.

ryanolsonk commented:
Wow! Very cool. I'll review this one soon.

ryanolsonk commented on Classes/GlobalStateExplorers/DatabaseBrowser/FLEXDatabaseManager.m:
```diff
@@ -0,0 +1,194 @@
+  
+  if (self) {
+    _databasePath = [aPath copy];
+    _db           = nil;
```
We can remove this line since all ivars are initialized with nil by default.

ryanolsonk commented on Classes/GlobalStateExplorers/DatabaseBrowser/FLEXDatabaseManager.m:
```diff
@@ -0,0 +1,194 @@
+  if (sqlite3_prepare_v2(_db, [sql UTF8String], -1, &pstmt, 0) == SQLITE_OK) {
+    while (sqlite3_step(pstmt) == SQLITE_ROW) {
+      NSUInteger num_cols = (NSUInteger)sqlite3_data_count(pstmt);
+      if (num_cols > 0) {
```
`num_cols` will always be `> 0` because it is unsigned. Maybe make it an `int` since that's what `sqlite3_data_count()` returns. Small nit on style too: prefer camel case instead of underscores for variable names.

ryanolsonk commented on Classes/GlobalStateExplorers/DatabaseBrowser/FLEXDatabaseManager.m:
```diff
@@ -0,0 +1,194 @@
+}
+
+
+- (BOOL)open {
```
Is this pulled from FMDB or somewhere else? That's probably ok given the licensing from FMDB should permit this use, but we should definitely add a note in the Thanks & Credits section of the README, and maybe a comment around code that was copied with a link back to the original source.

ryanolsonk commented on Classes/GlobalStateExplorers/DatabaseBrowser/FLEXMultiColumnTableView.h:
```diff
@@ -0,0 +1,47 @@
+
+@protocol FLEXMultiColumnTableViewDelegate <NSObject>
+
+@optional
```
Can we make these required instead of optional? That way we get more compile-time safety and don't need `respondsToSelector:` checks.

Small naming nits - I think these would be more consistent with Cocoa style patterns:
`-multiColumnTableView:didTapLabelWithText:`
`-multiColumnTableView:didTapHeaderWithText:sortType:`

ryanolsonk commented on Classes/GlobalStateExplorers/DatabaseBrowser/FLEXMultiColumnTableView.h:
```diff
@@ -0,0 +1,47 @@
+- (CGFloat)multiColumnTableView:(FLEXMultiColumnTableView *)tableView widthForContentCellInColumn:(NSInteger)column;
+- (CGFloat)multiColumnTableView:(FLEXMultiColumnTableView *)tableView heightForContentCellInRow:(NSInteger)row;
+- (CGFloat)heightForTopHeaderInTableView:(FLEXMultiColumnTableView *)tableView;
+- (CGFloat)WidthForLeftHeaderInTableView:(FLEXMultiColumnTableView *)tableView;
```
Style nit: W should be lowercase here.

ryanolsonk commented on Classes/GlobalStateExplorers/DatabaseBrowser/FLEXMultiColumnTableView.m:
```diff
@@ -0,0 +1,345 @@
+@interface FLEXMultiColumnTableView ()
+<UITableViewDataSource, UITableViewDelegate,UIScrollViewDelegate, FLEXTableContentCellDelegate>
+
+@property (nonatomic, weak) UIScrollView *contentScrollView;
```
What's the reasoning behind making these properties `weak`?

ryanolsonk commented on Classes/GlobalStateExplorers/DatabaseBrowser/FLEXTableColumnHeader.m:
```diff
@@ -0,0 +1,62 @@
+    case FLEXTableColumnHeaderSortTypeDesc:
+      _arrowLabel.text = @"⬇️";
+      break;
+    default:
```
We can kill the default statement here since we cover all the cases.

ryanolsonk commented on Classes/GlobalStateExplorers/DatabaseBrowser/FLEXTableColumnHeader.h:
```diff
@@ -0,0 +1,24 @@
+
+#import <UIKit/UIKit.h>
+
+typedef NS_ENUM(NSUInteger, FLEXTableColumnHeaderSortType) {
```
Nice :+1:

ryanolsonk commented on Classes/GlobalStateExplorers/DatabaseBrowser/FLEXTableContentViewController.h:
```diff
@@ -0,0 +1,16 @@
+@interface FLEXTableContentViewController : UIViewController
+
+@property (nonatomic, strong) NSArray *columnsArray;
+@property (nonatomic, strong) NSArray *contensArray;
```
`contentsArray`

ryanolsonk commented on Classes/GlobalStateExplorers/DatabaseBrowser/FLEXTableContentViewController.m:
```diff
@@ -0,0 +1,183 @@
+    
+    CGRect rectStatus = [UIApplication sharedApplication].statusBarFrame;
+    CGFloat y = 64;
+    if (rectStatus.size.height == 0) {
```
Not sure I understand what's going on here. Can you explain where the 32/64 numbers come from?

tttpeng replied:
If iPhone become landscape,The status bar is hidden,The height of navigation bar is 32 points, So the y = 32.

ryanolsonk commented on Classes/GlobalStateExplorers/DatabaseBrowser/FLEXTableContentViewController.m:
```diff
@@ -0,0 +1,183 @@
+#pragma mark -
+#pragma mark About Transition
+
+- (void)willTransitionToTraitCollection:(UITraitCollection *)newCollection
```
Do we need this since we have an autoresizing mask set on the tableview?

ryanolsonk commented on Classes/GlobalStateExplorers/DatabaseBrowser/FLEXTableContentViewController.m:
```diff
@@ -0,0 +1,183 @@
+  NSArray *sortContentData = [self.contensArray sortedArrayUsingComparator:^NSComparisonResult(id obj1, id obj2) {
+    
+    if ([obj1 objectForKey:text] == [NSNull null]) {
+      return 0;
```
Can we use the `NSComparisonResult` enum values here? Also, 0 corresponds to `NSOrderedSame` - I'm not sure that's what we want here.

ryanolsonk commented on Classes/GlobalStateExplorers/DatabaseBrowser/FLEXTableListViewController.m:
```diff
@@ -0,0 +1,87 @@
+  return self;
+}
+
+- (void)viewDidLoad
```
Empty override

ryanolsonk commented:
Nice work! This is really cool!

I left some comments on the changes in-line. My biggest concern is that we properly attribute any code that was copied from a 3rd party and make sure the licensing is compatible. There are some style inconsistencies with the rest of the project (e.g. 2 space indenting instead of 4), but I can just run clang-format over the changes after merging. Let me know once you've had a chance to look at the feedback.

Thanks again for making this pull request. This is going to be an awesome new feature for FLEX!

tttpeng commented:
Thanks for review my code so carefully, I learned a lot.

This is my first time to participate in open source projects,Forgive me have some not good enough,I corrected some mistakes you mentioned,I copied some code from FMDB,I see FMDB license is MIT LICENSE,I think I can do this,I added some explanation in the file,I do not know whether I'm doing right,If I do what is wrong. Please help me to correct,Because my English is poor,Thank you!

It is a valuable experience for me,Thank you.

ryanolsonk commented:
Looks great! Sorry for the delay on my part. Merging now. Thanks again for the contribution and for making the changes!

Cooooper commented:
The feature is Very nice! thanks!
